File: app/src/main/java/com/georgcantor/wallpaperapp/ui/adapter/FavoriteAdapter.java
package com.georgcantor.wallpaperapp.ui.adapter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.georgcantor.wallpaperapp.R;
import com.georgcantor.wallpaperapp.model.db.DatabaseHelper;
import com.georgcantor.wallpaperapp.model.db.Favorite;
import com.georgcantor.wallpaperapp.ui.FavDetailActivity;
import com.squareup.picasso.Picasso;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class FavoriteAdapter extends BaseAdapter {

    private Context context;
    private int layout;
    private ArrayList<Favorite> favoriteArrayList;
    private DatabaseHelper db;

    public FavoriteAdapter(Context context, int layout, ArrayList<Favorite> favoriteArrayList) {
        this.context = context;
        this.layout = layout;
        this.favoriteArrayList = favoriteArrayList;
    }

    @Override
    public int getCount() {
        return favoriteArrayList.size();
    }

    @Override
    public Object getItem(int position) {
        return favoriteArrayList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    private class ViewHolder {
        ImageView imageView;
        TextView textView;
    }

    @Override
    public View getView(final int position, final View view, ViewGroup viewGroup) {
        View row = view;
        ViewHolder holder = new ViewHolder();

        if (row == null) {
            LayoutInflater inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            row = inflater.inflate(layout, null);

            holder.textView = row.findViewById(R.id.timestamp);
            holder.imageView = row.findViewById(R.id.imgFavorite);
            row.setTag(holder);
        } else {
            holder = (ViewHolder) row.getTag();
        }

        Favorite favorite = favoriteArrayList.get(position);
        holder.textView.setText(formatDate(favorite.getTimestamp()));
        Picasso.with(context)
                .load(favorite.getImageUrl())
                .placeholder(R.drawable.plh)
                .into(holder.imageView);

        holder.imageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Activity activity = (Activity) context;
                Favorite photo = favoriteArrayList.get(position);
                final String url = photo.getImageUrl();
                final String hdUrl = photo.getHdUrl();
                Intent intent = new Intent(context, FavDetailActivity.class);
                intent.putExtra("preview", url);
                intent.putExtra("hd", hdUrl);
                context.startActivity(intent);
                activity.overridePendingTransition(R.anim.pull_in_right, R.anim.push_out_left);
            }
        });

        return row;
    }

    private String formatDate(String dateStr) {
        try {
            SimpleDateFormat fmt = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            Date date = fmt.parse(dateStr);
            SimpleDateFormat fmtOut = new SimpleDateFormat("MMM d");
            return fmtOut.format(date);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }
}

File: app/src/main/java/com/georgcantor/wallpaperapp/ui/adapter/FavoriteAdapter.kt
package com.georgcantor.wallpaperapp.ui.adapter

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.georgcantor.wallpaperapp.R
import com.georgcantor.wallpaperapp.model.CommonPic
import com.georgcantor.wallpaperapp.model.local.db.DatabaseHelper
import com.georgcantor.wallpaperapp.model.local.db.Favorite
import com.georgcantor.wallpaperapp.ui.DetailsActivity
import com.georgcantor.wallpaperapp.ui.adapter.holder.FavoriteViewHolder
import com.georgcantor.wallpaperapp.util.longToast
import com.georgcantor.wallpaperapp.util.showDialog
import com.google.gson.Gson
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import java.util.*
import java.util.concurrent.TimeUnit

class FavoriteAdapter(private val context: Context) : RecyclerView.Adapter<FavoriteViewHolder>() {

    private val favorites: MutableList<Favorite>?
    private val activity = context as Activity
    private val db = DatabaseHelper(context)

    init {
        this.favorites = ArrayList()
    }

    fun setFavList(strings: MutableList<Favorite>) {
        clearPicList()
        this.favorites?.addAll(strings)
        notifyDataSetChanged()
    }

    private fun clearPicList() {
        val size = favorites?.size
        favorites?.clear()
        size?.let { notifyItemRangeRemoved(0, it) }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FavoriteViewHolder {
        val itemView = LayoutInflater.from(context).inflate(R.layout.favorite_item, null)

        return FavoriteViewHolder(itemView)
    }

    @SuppressLint("CheckResult")
    override fun onBindViewHolder(holder: FavoriteViewHolder, position: Int) {
        val favorite = favorites?.get(position)
        val hitJson = favorite?.hit

        val gson = Gson()
        val pic = gson.fromJson(hitJson, CommonPic::class.java)

        val publishSubject = PublishSubject.create<Int>()
        publishSubject
            .throttleFirst(1, TimeUnit.SECONDS)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                val intent = Intent(context, DetailsActivity::class.java)
                try {
                    intent.putExtra(
                        DetailsActivity.EXTRA_PIC,
                        CommonPic(
                            url = pic.url,
                            width = pic.width,
                            heght = pic.heght,
                            likes = pic.likes,
                            favorites = pic.favorites,
                            tags = pic.tags,
                            downloads = pic.downloads,
                            imageURL = pic.imageURL,
                            fullHDURL = pic.fullHDURL,
                            user = pic.user,
                            userImageURL = pic.userImageURL
                        )
                    )
                } catch (e: ArrayIndexOutOfBoundsException) {
                    context.longToast(context.getString(R.string.something_went_wrong))
                }
                context.startActivity(intent)
                activity.overridePendingTransition(R.anim.pull_in_right, R.anim.push_out_left)
            }

        holder.imageView.setOnClickListener {
            publishSubject.onNext(0)
        }

        holder.imageView.setOnLongClickListener {
            val photo = favorites?.get(position)
            val url = photo?.imageUrl

            context.showDialog(context.getString(R.string.del_from_fav_dialog)) {
                deleteFromFavorites(url)
            }
            false
        }

        val layoutParams = holder.imageView.layoutParams as RelativeLayout.LayoutParams
        val height = pic.heght.toFloat()
        val width = pic.width.toFloat()
        val ratio = height.div(width)
        layoutParams.height = (layoutParams.width * ratio).toInt()
        holder.imageView.layoutParams = layoutParams
        holder.imageView.setRatio(ratio)

        Glide.with(context)
            .load(pic.url)
            .placeholder(R.drawable.plh)
            .thumbnail(0.1f)
            .into(holder.imageView)
    }

    override fun getItemCount(): Int = favorites?.size ?: 0

    private fun deleteFromFavorites(url: String?) {
        if (url != null) {
            db.deleteFromFavorites(url)
        }
        activity.recreate()
    }

}